lib/evm/operation.ex

```diff
@@ -0,0 +1,340 @@
+  def encode(operation) when is_atom(operation) do
+    Map.get(@operations_to_opcodes, operation)
+  end
+
+  @doc """
+  Returns metadata about a given operation or opcode, or nil.
+
+  ## Examples
+
+      iex> EVM.Operation.metadata(:stop)
+      %EVM.Operation.Metadata{id: 0x00, sym: :stop, input_count: 0, output_count: 0, description: "Halts execution", group: :stop_and_arithmetic}
+
+      iex> EVM.Operation.metadata(0x00)
+      %EVM.Operation.Metadata{id: 0x00, sym: :stop, input_count: 0, output_count: 0, description: "Halts execution", group: :stop_and_arithmetic}
+
+      iex> EVM.Operation.metadata(:add)
+      %EVM.Operation.Metadata{id: 0x01, sym: :add, input_count: 2, output_count: 1, description: "Addition operation", group: :stop_and_arithmetic}
+
+      iex> EVM.Operation.metadata(:push1)
+      %EVM.Operation.Metadata{id: 0x60, sym: :push1, fun: :push_n, args: [1], input_count: 0, output_count: 1, description: "Place 1-byte item on stack", group: :push}
+
+      iex> EVM.Operation.metadata(0xfe)
+      nil
+
+      iex> EVM.Operation.metadata(nil)
+      nil
+  """
+  @spec metadata(operation | opcode) :: EVM.Operation.Metadata.t | nil
+  def metadata(nil), do: nil
+  def metadata(operation) when is_atom(operation) do
+    operation |> encode |> metadata
+  end
+
+  def metadata(opcode) when is_integer(opcode) do
+    Map.get(@opcodes_to_metadata, opcode)
+  end
+
+  @doc """
+  Executes a single operation. This simply does the effects of the operation itself,
+  ignoring the rest of the actions of an operation cycle. This will effect, for instance,
+  the stack, but will not effect the gas, etc.
+
+  ## Examples
+
+      # TODO: How to handle trie state in tests?
+
+      # Add
+      iex> EVM.Operation.run_operation(:add, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{})
+      {%{}, %EVM.MachineState{stack: [3]}, %EVM.SubState{}, %EVM.ExecEnv{}}
+
+      # Push
+      iex> EVM.Operation.run_operation(:push1, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{machine_code: <<00, 01>>})
+      {%{}, %EVM.MachineState{stack: [1, 1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{machine_code: <<00, 01>>}}
+
+      # nil
+      iex> EVM.Operation.run_operation(:stop, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{})
+      {%{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{}}
+
+      # Unimplemented
+      iex> EVM.Operation.run_operation(:log0, %{}, %EVM.MachineState{stack: [1, 2]}, %EVM.SubState{}, %EVM.ExecEnv{})
+      {%{}, %EVM.MachineState{stack: []}, %EVM.SubState{}, %EVM.ExecEnv{}}
+  """
+  @spec run_operation(operation, EVM.state, MachineState.t, SubState.t, ExecEnv.t) :: {EVM.state, MachineState.t, SubState.t, ExecEnv.t}
+  def run_operation(operation, state, machine_state, sub_state, exec_env) do
+    {args, updated_stack, updated_machine_state} = operation_args(operation, state, machine_state, sub_state, exec_env)
```

Comment:
This should probably just return `{args, machine_state}` and you can just refer to stack as `updated_machine_state.stack` below.